[BUG] @prizm-ui/components - Switcher не реагирует на выставление выбранного элемента если опции приходят асинхронно

### Библиотека

_Выберите нужное отметив галочку [x]_

- [ ] `@prizm-ui/core`
- [x] `@prizm-ui/components`
- [ ] `@prizm-ui/install`
- [ ] `@prizm-ui/icons`
- [ ] `@prizm-ui/flag-icons`
- [ ] `@prizm-ui/theme`
- [ ] `@prizm-ui/charts`
- [ ] `@prizm-ui/ast`
- [ ] `@prizm-ui/nx-plugin`
-

### Компонент

prizm-switcher

### Версия

Операционная ситема:

Windows 10

Браузер: Chrome Version 122.0.6261.129 (Official Build) (64-bit)

`Google Chrome Версия 96.0.4664.110 (Официальная сборка), (x86_64)`

NodeJS:

```
➜ node --version
v16.13.0
```

Dependencies:

```
➜ npm ls @prizm-ui/components

@prizm-ui/components 1.4.4
```

```
➜ npm ls @angular/core
$ npm ls @angular/core
npm ERR! code ELSPROBLEMS
npm ERR! invalid: @angular/core@14.2.12 C:\projects\Energy Control\front\node_modules\@angular\core
ec-front@0.0.0 C:\projects\Energy Control\front
├─┬ @angular/animations@14.2.12
│ └── @angular/core@14.2.12 deduped
├─┬ @angular/cdk@14.2.7
│ └── @angular/core@14.2.12 deduped
├─┬ @angular/common@14.2.12
│ └── @angular/core@14.2.12 deduped
├─┬ @angular/compiler@14.2.12
│ └── @angular/core@14.2.12 deduped
├── @angular/core@14.2.12
├─┬ @angular/forms@14.2.12
│ └── @angular/core@14.2.12 deduped
├─┬ @angular/material@14.2.7
│ └── @angular/core@14.2.12 deduped
├─┬ @angular/platform-browser-dynamic@14.2.12
│ └── @angular/core@14.2.12 deduped
├─┬ @angular/platform-browser@14.2.12
│ └── @angular/core@14.2.12 deduped
├─┬ @angular/router@14.2.12
│ └── @angular/core@14.2.12 deduped
├─┬ @ng-web-apis/common@1.13.0
│ └── @angular/core@14.2.12 deduped
├─┬ @ng-web-apis/intersection-observer@3.0.0
│ └── @angular/core@14.2.12 deduped
├─┬ @ng-web-apis/resize-observer@2.0.0
│ └── @angular/core@14.2.12 deduped
├─┬ @ngrx/component-store@14.3.3
│ └── @angular/core@14.2.12 deduped
├─┬ @ngrx/effects@14.3.2
│ └── @angular/core@14.2.12 deduped
├─┬ @ngrx/entity@14.3.2
│ └── @angular/core@14.2.12 deduped
├─┬ @ngrx/router-store@14.3.2
│ └── @angular/core@14.2.12 deduped
├─┬ @ngrx/store@14.3.2
│ └── @angular/core@14.2.12 deduped
├─┬ @prizm-ui/components@1.4.4
│ ├── @angular/core@14.2.12 deduped
│ └─┬ @ng-web-apis/mutation-observer@2.0.0
│   └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon, "^7.0.0 || ^8.0.0 || ^9.0.0 || ^10.0.0 || ^11.0.0 || ^12.0.0" from node_modules/primeng
├─┬ @prizm-ui/core@1.4.4
│ └── @angular/core@14.2.12 deduped
├─┬ @prizm-ui/helpers@1.4.4
│ └── @angular/core@14.2.12 deduped
├─┬ @prizm-ui/i18n@1.4.4
│ └── @angular/core@14.2.12 deduped
├─┬ @prizm-ui/icons@1.4.4
│ └── @angular/core@14.2.12 deduped
├─┬ @prizm-ui/theme@1.4.4
│ └── @angular/core@14.2.12 deduped
├─┬ @taiga-ui/cdk@3.15.0
│ ├── @angular/core@14.2.12 deduped
│ ├─┬ @ng-web-apis/common@2.0.1
│ │ └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon, "^7.0.0 || ^8.0.0 || ^9.0.0 || ^10.0.0 || ^11.0.0 || ^12.0.0" from node_modules/primeng
│ ├─┬ @tinkoff/ng-event-plugins@3.1.0
│ │ └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon, "^7.0.0 || ^8.0.0 || ^9.0.0 || ^10.0.0 || ^11.0.0 || ^12.0.0" from node_modules/primeng
│ └─┬ @tinkoff/ng-polymorpheus@4.0.10
│   └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon, "^7.0.0 || ^8.0.0 || ^9.0.0 || ^10.0.0 || ^11.0.0 || ^12.0.0" from node_modules/primeng
├─┬ angular-svg-icon@15.0.0
│ └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon
├─┬ jest-preset-angular@13.1.2
│ └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon
├─┬ keycloak-angular@12.2.0
│ └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon
├─┬ ngx-mask@12.0.0 invalid: "14.2.4" from node_modules/@prizm-ui/components
│ └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon
└─┬ primeng@12.2.3
  └── @angular/core@14.2.12 deduped invalid: ">=15.0.0" from node_modules/angular-svg-icon, "^7.0.0 || ^8.0.0 || ^9.0.0 || ^10.0.0 || ^11.0.0 || ^12.0.0" from node_modules/primeng

```

### Проект в котором используется Prizm

- Энергоконтроль

### Скрин / Видео проблемы



### Воспроизведение проблемы

Проблема воспроизводится, когда мы используем prizm-switcher, опции для которого "switchers"  устанавливаются через пайп async. Если выставить желаемое значение для выбора, до того как, опции пришли, когда опции отрисуются значение не будет выбрано а выбрана будет первая опция.

```
<prizm-switcher [switchers]="switchers$ | async" [selectedSwitcherIdx]="1" type="inner" size="l"></prizm-switcher>
```
Ожидаем что будет выбрана вторя опция, но выберется первая.

Так же это актруально если просто задать индекс выбранног элемента раньше чем опции
```
<prizm-switcher [selectedSwitcherIdx]="1" [switchers]="[{title: '1'},{title: '2'}]"  type="inner" size="l"></prizm-switcher>
```
Ожидаем что будет выбрана вторя опция, но выберется первая.

### Описание проблемы

--- a/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.html
+++ b/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.html
@@ -1,0 +1,13 @@
+<prizm-switcher [switchers]="switchers" [selectedSwitcherIdx]="1"></prizm-switcher>
+
+<br />
+<br />
+
+<prizm-switcher [switchers]="(switchers$ | async)!" [selectedSwitcherIdx]="1"></prizm-switcher>
+
+<br />
+<br />
+
+<!--  Note: properties order matters  -->
+
+<prizm-switcher [selectedSwitcherIdx]="1" [switchers]="switchers"></prizm-switcher>

--- a/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.html
+++ b/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.html
@@ -1,13 +1,25 @@
-<prizm-switcher [switchers]="switchers" [selectedSwitcherIdx]="1"></prizm-switcher>
+<h2>Static</h2>
+
+<prizm-switcher [switchers]="switchers" [selectedSwitcherIdx]="selectedIndex"></prizm-switcher>
 
 <br />
 <br />
 
-<prizm-switcher [switchers]="(switchers$ | async)!" [selectedSwitcherIdx]="1"></prizm-switcher>
+<h2>Async</h2>
+<prizm-switcher [switchers]="(switchers$ | async)!" [selectedSwitcherIdx]="selectedIndex"></prizm-switcher>
 
 <br />
 <br />
 
-<!--  Note: properties order matters  -->
+<h2>Inverse attribute ordering</h2>
 
-<prizm-switcher [selectedSwitcherIdx]="1" [switchers]="switchers"></prizm-switcher>
+<prizm-switcher [selectedSwitcherIdx]="selectedIndex" [switchers]="switchers"></prizm-switcher>
+
+<br />
+<br />
+
+<button class="margin-right" (click)="toggleIndex()" prizmButton>
+  Update index. Current index is {{ selectedIndex }}
+</button>
+
+<button (click)="toggleSwitchers()" prizmButton>Update switchers</button>

--- a/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.html
+++ b/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.html
@@ -1,19 +1,31 @@
 <h2>Static</h2>
 
-<prizm-switcher [switchers]="switchers" [selectedSwitcherIdx]="selectedIndex"></prizm-switcher>
+<prizm-switcher
+  [switchers]="switchers"
+  [selectedSwitcherIdx]="selectedIndex"
+  (selectedSwitcherIdxChange)="updateIdx($event)"
+></prizm-switcher>
 
 <br />
 <br />
 
 <h2>Async</h2>
-<prizm-switcher [switchers]="(switchers$ | async)!" [selectedSwitcherIdx]="selectedIndex"></prizm-switcher>
+<prizm-switcher
+  [switchers]="(switchers$ | async)!"
+  [selectedSwitcherIdx]="selectedIndex"
+  (selectedSwitcherIdxChange)="updateIdx($event)"
+></prizm-switcher>
 
 <br />
 <br />
 
 <h2>Inverse attribute ordering</h2>
 
-<prizm-switcher [selectedSwitcherIdx]="selectedIndex" [switchers]="switchers"></prizm-switcher>
+<prizm-switcher
+  [selectedSwitcherIdx]="selectedIndex"
+  [switchers]="switchers"
+  (selectedSwitcherIdxChange)="updateIdx($event)"
+></prizm-switcher>
 
 <br />
 <br />

--- a/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.less
+++ b/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.less
@@ -1,0 +1,3 @@
+.margin-right {
+  margin-right: 24px;
+}

--- a/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.ts
+++ b/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.ts
@@ -12,6 +12,7 @@ import { BehaviorSubject, switchMap, takeUntil, tap, timer } from 'rxjs';
 })
 export class SwitcherAsyncExampleComponent {
   public selectedIndex = 1;
+  public testIndex = this.selectedIndex;
   public readonly switchersSetA: PrizmSwitcherItem[] = [
     {
       title: 'Таблицы',
@@ -59,5 +60,9 @@ export class SwitcherAsyncExampleComponent {
     this.switchers === this.switchersSetA
       ? this.switchers$$.next(this.switchersSetB)
       : this.switchers$$.next(this.switchersSetA);
+  }
+
+  public updateIdx(idx: number) {
+    this.selectedIndex = idx;
   }
 }

--- a/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.ts
+++ b/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.ts
@@ -1,0 +1,34 @@
+import { Component, ChangeDetectionStrategy } from '@angular/core';
+import { PrizmSwitcherItem } from '@prizm-ui/components';
+import { PrizmDestroyService } from '@prizm-ui/helpers';
+import { map, takeUntil, timer } from 'rxjs';
+
+@Component({
+  selector: 'prizm-switcher-async-example',
+  templateUrl: './switcher-async-example.component.html',
+  styleUrls: ['./switcher-async-example.component.less'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  providers: [PrizmDestroyService],
+})
+export class SwitcherAsyncExampleComponent {
+  public readonly switchers: PrizmSwitcherItem[] = [
+    {
+      title: 'Таблицы',
+    },
+    {
+      title: 'Графики',
+    },
+    {
+      title: 'Мнемосхемы',
+    },
+    {
+      title: 'Дашборды',
+    },
+  ];
+
+  public readonly switchers$ = timer(100).pipe(
+    map(() => this.switchers),
+    takeUntil(this.destroy$)
+  );
+  constructor(private readonly destroy$: PrizmDestroyService) {}
+}

--- a/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.ts
+++ b/apps/doc/src/app/components/switcher/examples/switcher-async-example/switcher-async-example.component.ts
@@ -1,7 +1,7 @@
 import { Component, ChangeDetectionStrategy } from '@angular/core';
 import { PrizmSwitcherItem } from '@prizm-ui/components';
 import { PrizmDestroyService } from '@prizm-ui/helpers';
-import { map, takeUntil, timer } from 'rxjs';
+import { BehaviorSubject, switchMap, takeUntil, tap, timer } from 'rxjs';
 
 @Component({
   selector: 'prizm-switcher-async-example',
@@ -11,7 +11,8 @@ import { map, takeUntil, timer } from 'rxjs';
   providers: [PrizmDestroyService],
 })
 export class SwitcherAsyncExampleComponent {
-  public readonly switchers: PrizmSwitcherItem[] = [
+  public selectedIndex = 1;
+  public readonly switchersSetA: PrizmSwitcherItem[] = [
     {
       title: 'Таблицы',
     },
@@ -20,15 +21,43 @@ export class SwitcherAsyncExampleComponent {
     },
     {
       title: 'Мнемосхемы',
+      disabled: true,
     },
     {
       title: 'Дашборды',
     },
   ];
 
+  public readonly switchersSetB: PrizmSwitcherItem[] = [
+    {
+      title: 'Москва',
+    },
+    {
+      title: 'Санкт-Петербург',
+    },
+  ];
+
+  public switchers: PrizmSwitcherItem[] = this.switchersSetA;
+
   public readonly switchers$ = timer(100).pipe(
-    map(() => this.switchers),
+    switchMap(() => this.switchers$$),
+    tap(value => (this.switchers = value)),
     takeUntil(this.destroy$)
   );
+
+  private readonly switchers$$: BehaviorSubject<PrizmSwitcherItem[]> = new BehaviorSubject<
+    PrizmSwitcherItem[]
+  >(this.switchersSetA);
+
   constructor(private readonly destroy$: PrizmDestroyService) {}
+
+  public toggleIndex(): void {
+    this.selectedIndex === 1 ? (this.selectedIndex = 2) : (this.selectedIndex = 1);
+  }
+
+  public toggleSwitchers(): void {
+    this.switchers === this.switchersSetA
+      ? this.switchers$$.next(this.switchersSetB)
+      : this.switchers$$.next(this.switchersSetA);
+  }
 }

--- a/apps/doc/src/app/components/switcher/switcher-example.component.html
+++ b/apps/doc/src/app/components/switcher/switcher-example.component.html
@@ -7,6 +7,10 @@
       <prizm-switcher-basic-example></prizm-switcher-basic-example>
     </prizm-doc-example>
 
+    <prizm-doc-example id="async" [content]="exampleAsyncSwitcher" heading="Async">
+      <prizm-switcher-async-example></prizm-switcher-async-example>
+    </prizm-doc-example>
+
     <prizm-doc-example id="innerL" [content]="exampleInnerLSwitcher" heading="Type inner, size L">
       <prizm-switcher-inner-l-example></prizm-switcher-inner-l-example>
     </prizm-doc-example>

--- a/apps/doc/src/app/components/switcher/switcher-example.component.ts
+++ b/apps/doc/src/app/components/switcher/switcher-example.component.ts
@@ -39,6 +39,11 @@ export class SwitcherExampleComponent {
     HTML: import('./examples/switcher-basic-example/switcher-basic-example.component.html?raw'),
   };
 
+  public readonly exampleAsyncSwitcher: TuiDocExample = {
+    TypeScript: import('./examples/switcher-async-example/switcher-async-example.component?raw'),
+    HTML: import('./examples/switcher-async-example/switcher-async-example.component.html?raw'),
+  };
+
   public readonly exampleInnerLSwitcher: TuiDocExample = {
     TypeScript: import('./examples/switcher-inner-l-example/switcher-inner-l-example.component?raw'),
     HTML: import('./examples/switcher-inner-l-example/switcher-inner-l-example.component.html?raw'),

--- a/apps/doc/src/app/components/switcher/switcher-example.module.ts
+++ b/apps/doc/src/app/components/switcher/switcher-example.module.ts
@@ -4,7 +4,7 @@ import { SwitcherExampleComponent } from './switcher-example.component';
 import { prizmDocGenerateRoutes, PrizmAddonDocModule } from '@prizm-ui/doc';
 import { RouterModule } from '@angular/router';
 import { SwitcherBasicExampleComponent } from './examples/switcher-basic-example/switcher-basic-example.component';
-import { PrizmSwitcherModule } from '@prizm-ui/components';
+import { PrizmButtonComponent, PrizmSwitcherModule } from '@prizm-ui/components';
 import { SwitcherInnerLExampleComponent } from './examples/switcher-inner-l-example/switcher-inner-l-example.component';
 import { SwitcherInnerMExampleComponent } from './examples/switcher-inner-m-example/switcher-inner-m-example.component';
 import { SwitcherOuterMExampleComponent } from './examples/switcher-outer-m-example/switcher-outer-m-example.component';
@@ -34,6 +34,7 @@ import { SwitcherAsyncExampleComponent } from './examples/switcher-async-example
     RouterModule.forChild(prizmDocGenerateRoutes(SwitcherExampleComponent)),
     PrizmSwitcherModule,
     ReactiveFormsModule,
+    PrizmButtonComponent,
   ],
 })
 export class SwitcherExampleModule {}

--- a/apps/doc/src/app/components/switcher/switcher-example.module.ts
+++ b/apps/doc/src/app/components/switcher/switcher-example.module.ts
@@ -13,6 +13,7 @@ import { SwitcherOuterSExampleComponent } from './examples/switcher-outer-s-exam
 import { SwitcherWithIconExampleComponent } from './examples/switcher-with-icon-example/switcher-with-icon-example.component';
 import { SwitcherOnlyIconExampleComponent } from './examples/switcher-only-icon-example/switcher-only-icon-example.component';
 import { ReactiveFormsModule } from '@angular/forms';
+import { SwitcherAsyncExampleComponent } from './examples/switcher-async-example/switcher-async-example.component';
 
 @NgModule({
   declarations: [
@@ -25,6 +26,7 @@ import { ReactiveFormsModule } from '@angular/forms';
     SwitcherOuterSExampleComponent,
     SwitcherWithIconExampleComponent,
     SwitcherOnlyIconExampleComponent,
+    SwitcherAsyncExampleComponent,
   ],
   imports: [
     CommonModule,

--- a/libs/components/src/lib/components/switcher/switcher.component.ts
+++ b/libs/components/src/lib/components/switcher/switcher.component.ts
@@ -5,22 +5,21 @@ import {
   EventEmitter,
   HostBinding,
   Input,
-  OnChanges,
+  OnInit,
   Optional,
   Output,
   Self,
-  SimpleChanges,
 } from '@angular/core';
 import { PrizmSwitcherItem, PrizmSwitcherSize, PrizmSwitcherType } from './switcher.interface';
 import { prizmDefaultProp } from '@prizm-ui/core';
 import { PrizmAbstractTestId } from '../../abstract/interactive';
 import { ControlValueAccessor, NgControl } from '@angular/forms';
-import { noop } from 'rxjs';
+import { BehaviorSubject, combineLatestWith, noop, takeUntil, tap } from 'rxjs';
 import { CommonModule } from '@angular/common';
 import { PrizmSwitcherItemComponent } from './components/switcher-item/switcher-item.component';
 import { PrizmSwitcherHintDirective } from './directives/switcher-hint.directive';
-import { prizmHasChanges } from '@prizm-ui/helpers';
 import { INITIAL_SWITHCER_INDEX } from './swithcer.const';
+import { PrizmDestroyService } from '@prizm-ui/helpers';
 
 @Component({
   selector: 'prizm-switcher',
@@ -29,8 +28,9 @@ import { INITIAL_SWITHCER_INDEX } from './swithcer.const';
   changeDetection: ChangeDetectionStrategy.OnPush,
   standalone: true,
   imports: [CommonModule, PrizmSwitcherHintDirective, PrizmSwitcherItemComponent],
+  providers: [PrizmDestroyService],
 })
-export class PrizmSwitcherComponent extends PrizmAbstractTestId implements ControlValueAccessor, OnChanges {
+export class PrizmSwitcherComponent extends PrizmAbstractTestId implements ControlValueAccessor, OnInit {
   @Input()
   @prizmDefaultProp()
   public size: PrizmSwitcherSize = 'l';
@@ -39,13 +39,28 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
   @prizmDefaultProp()
   public type: PrizmSwitcherType = 'inner';
 
-  @Input()
-  @prizmDefaultProp()
-  public switchers: PrizmSwitcherItem[] = [];
+  private swithcers$: BehaviorSubject<PrizmSwitcherItem[]> = new BehaviorSubject<PrizmSwitcherItem[]>([]);
 
   @Input()
   @prizmDefaultProp()
-  public selectedSwitcherIdx: number = INITIAL_SWITHCER_INDEX;
+  public set switchers(value: PrizmSwitcherItem[]) {
+    if (value) this.swithcers$.next(value);
+  }
+  get switchers(): PrizmSwitcherItem[] {
+    return this.swithcers$.value;
+  }
+
+  private selectedSwitcherIdx$: BehaviorSubject<number> = new BehaviorSubject(INITIAL_SWITHCER_INDEX);
+  public selectedSwitcherIdx_ = INITIAL_SWITHCER_INDEX;
+
+  @Input()
+  @prizmDefaultProp()
+  public set selectedSwitcherIdx(value: number) {
+    this.selectedSwitcherIdx$.next(value);
+  }
+  get selectedSwitcherIdx(): number {
+    return this.selectedSwitcherIdx_;
+  }
 
   @Input()
   @HostBinding('class.full-width')
@@ -61,18 +76,19 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
 
   constructor(
     public readonly cdRef: ChangeDetectorRef,
-    @Optional() @Self() public readonly ngControl: NgControl
+    @Optional() @Self() public readonly ngControl: NgControl,
+    private readonly destroy$: PrizmDestroyService
   ) {
     super();
     if (this.ngControl != null) {
       this.ngControl.valueAccessor = this;
     }
   }
-  ngOnChanges(changes: SimpleChanges): void {
-    if (prizmHasChanges(changes, 'selectedSwitcherIdx') && prizmHasChanges(changes, 'switchers')) {
-      this.setActiveSwitcherOnChange();
-    }
+
+  ngOnInit(): void {
+    this.handleSwitchersUpdate();
   }
+
   public selectSwitcher(item: PrizmSwitcherItem, idx: number): void {
     if (this.ngControl?.disabled) return;
     if (item.disabled) return;
@@ -95,10 +111,33 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
     this.cdRef.markForCheck();
   }
 
-  private setActiveSwitcherOnChange(): void {
-    const idx =
-      this.selectedSwitcherIdx <= this.switchers.length ? this.selectedSwitcherIdx : INITIAL_SWITHCER_INDEX;
-    const item = this.switchers[idx];
-    if (item) this.selectSwitcher(item, idx);
+  private handleSwitchersUpdate() {
+    this.swithcers$
+      .pipe(
+        tap(switchers => {
+          if (!this.isIndexValid(this.selectedSwitcherIdx_, switchers)) {
+            this.selectedSwitcherIdx_ = INITIAL_SWITHCER_INDEX;
+          }
+        }),
+        combineLatestWith(this.selectedSwitcherIdx$),
+        tap(([switchers, selectedSwitcherIdx]) => {
+          if (!switchers.length) return;
+
+          if (!this.isIndexValid(selectedSwitcherIdx, switchers)) {
+            console.warn('selectedSwitcherIdx out of bound');
+            return;
+          }
+
+          this.selectedSwitcherIdx_ = selectedSwitcherIdx;
+
+          this.selectSwitcher(switchers[selectedSwitcherIdx], selectedSwitcherIdx);
+        }),
+        takeUntil(this.destroy$)
+      )
+      .subscribe();
+  }
+
+  private isIndexValid(idx: number, switchers: PrizmSwitcherItem[]): boolean {
+    return !!switchers[idx];
   }
 }

--- a/libs/components/src/lib/components/switcher/switcher.component.ts
+++ b/libs/components/src/lib/components/switcher/switcher.component.ts
@@ -5,9 +5,11 @@ import {
   EventEmitter,
   HostBinding,
   Input,
+  OnChanges,
   Optional,
   Output,
   Self,
+  SimpleChanges,
 } from '@angular/core';
 import { PrizmSwitcherItem, PrizmSwitcherSize, PrizmSwitcherType } from './switcher.interface';
 import { prizmDefaultProp } from '@prizm-ui/core';
@@ -17,6 +19,8 @@ import { noop } from 'rxjs';
 import { CommonModule } from '@angular/common';
 import { PrizmSwitcherItemComponent } from './components/switcher-item/switcher-item.component';
 import { PrizmSwitcherHintDirective } from './directives/switcher-hint.directive';
+import { prizmHasChanges } from '@prizm-ui/helpers';
+import { INITIAL_SWITHCER_INDEX } from './swithcer.const';
 
 @Component({
   selector: 'prizm-switcher',
@@ -26,11 +30,7 @@ import { PrizmSwitcherHintDirective } from './directives/switcher-hint.directive
   standalone: true,
   imports: [CommonModule, PrizmSwitcherHintDirective, PrizmSwitcherItemComponent],
 })
-export class PrizmSwitcherComponent extends PrizmAbstractTestId implements ControlValueAccessor {
-  onChange: (v: number) => void = noop;
-  onTouched: () => void = noop;
-  private selectedSwitcherIdx_ = 0;
-
+export class PrizmSwitcherComponent extends PrizmAbstractTestId implements ControlValueAccessor, OnChanges {
   @Input()
   @prizmDefaultProp()
   public size: PrizmSwitcherSize = 'l';
@@ -45,13 +45,7 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
 
   @Input()
   @prizmDefaultProp()
-  public set selectedSwitcherIdx(idx: number) {
-    const item = this.switchers[idx];
-    if (item) this.selectSwitcher(item, idx);
-  }
-  get selectedSwitcherIdx() {
-    return this.selectedSwitcherIdx_;
-  }
+  public selectedSwitcherIdx: number = INITIAL_SWITHCER_INDEX;
 
   @Input()
   @HostBinding('class.full-width')
@@ -62,6 +56,9 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
 
   override readonly testId_ = 'ui_switcher';
 
+  onChange: (v: number) => void = noop;
+  onTouched: () => void = noop;
+
   constructor(
     public readonly cdRef: ChangeDetectorRef,
     @Optional() @Self() public readonly ngControl: NgControl
@@ -71,11 +68,19 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
       this.ngControl.valueAccessor = this;
     }
   }
+  ngOnChanges(changes: SimpleChanges): void {
+    if (prizmHasChanges(changes, 'selectedSwitcherIdx') && prizmHasChanges(changes, 'switchers')) {
+      const idx =
+        this.selectedSwitcherIdx <= this.switchers.length ? this.selectedSwitcherIdx : INITIAL_SWITHCER_INDEX;
+      const item = this.switchers[idx];
+      if (item) this.selectSwitcher(item, idx);
+    }
+  }
   public selectSwitcher(item: PrizmSwitcherItem, idx: number): void {
     if (this.ngControl?.disabled) return;
     if (item.disabled) return;
     if (this.selectedSwitcherIdx === idx) return;
-    this.selectedSwitcherIdxChange.emit((this.selectedSwitcherIdx_ = idx));
+    this.selectedSwitcherIdxChange.emit((this.selectedSwitcherIdx = idx));
     this.onChange(this.selectedSwitcherIdx);
   }
 

--- a/libs/components/src/lib/components/switcher/switcher.component.ts
+++ b/libs/components/src/lib/components/switcher/switcher.component.ts
@@ -70,10 +70,7 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
   }
   ngOnChanges(changes: SimpleChanges): void {
     if (prizmHasChanges(changes, 'selectedSwitcherIdx') && prizmHasChanges(changes, 'switchers')) {
-      const idx =
-        this.selectedSwitcherIdx <= this.switchers.length ? this.selectedSwitcherIdx : INITIAL_SWITHCER_INDEX;
-      const item = this.switchers[idx];
-      if (item) this.selectSwitcher(item, idx);
+      this.setActiveSwitcherOnChange();
     }
   }
   public selectSwitcher(item: PrizmSwitcherItem, idx: number): void {
@@ -96,5 +93,12 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
 
   public setDisabledState(isDisabled: boolean): void {
     this.cdRef.markForCheck();
+  }
+
+  private setActiveSwitcherOnChange(): void {
+    const idx =
+      this.selectedSwitcherIdx <= this.switchers.length ? this.selectedSwitcherIdx : INITIAL_SWITHCER_INDEX;
+    const item = this.switchers[idx];
+    if (item) this.selectSwitcher(item, idx);
   }
 }

--- a/libs/components/src/lib/components/switcher/switcher.component.ts
+++ b/libs/components/src/lib/components/switcher/switcher.component.ts
@@ -14,7 +14,7 @@ import { PrizmSwitcherItem, PrizmSwitcherSize, PrizmSwitcherType } from './switc
 import { prizmDefaultProp } from '@prizm-ui/core';
 import { PrizmAbstractTestId } from '../../abstract/interactive';
 import { ControlValueAccessor, NgControl } from '@angular/forms';
-import { BehaviorSubject, combineLatestWith, noop, takeUntil, tap } from 'rxjs';
+import { BehaviorSubject, combineLatestWith, filter, noop, takeUntil, tap } from 'rxjs';
 import { CommonModule } from '@angular/common';
 import { PrizmSwitcherItemComponent } from './components/switcher-item/switcher-item.component';
 import { PrizmSwitcherHintDirective } from './directives/switcher-hint.directive';
@@ -39,15 +39,15 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
   @prizmDefaultProp()
   public type: PrizmSwitcherType = 'inner';
 
-  private swithcers$: BehaviorSubject<PrizmSwitcherItem[]> = new BehaviorSubject<PrizmSwitcherItem[]>([]);
+  private switchers$: BehaviorSubject<PrizmSwitcherItem[]> = new BehaviorSubject<PrizmSwitcherItem[]>([]);
 
   @Input()
   @prizmDefaultProp()
   public set switchers(value: PrizmSwitcherItem[]) {
-    if (value) this.swithcers$.next(value);
+    if (value) this.switchers$.next(value);
   }
   get switchers(): PrizmSwitcherItem[] {
-    return this.swithcers$.value;
+    return this.switchers$.value;
   }
 
   private selectedSwitcherIdx$: BehaviorSubject<number> = new BehaviorSubject(INITIAL_SWITHCER_INDEX);
@@ -112,19 +112,21 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
   }
 
   private handleSwitchersUpdate() {
-    this.swithcers$
+    this.switchers$
       .pipe(
+        filter((switchers: PrizmSwitcherItem[]) => !!switchers.length),
         tap(switchers => {
           if (!this.isIndexValid(this.selectedSwitcherIdx_, switchers)) {
-            this.selectedSwitcherIdx_ = INITIAL_SWITHCER_INDEX;
+            this.selectSwitcher(switchers[INITIAL_SWITHCER_INDEX], INITIAL_SWITHCER_INDEX);
+            this.logIndexValidationError(
+              `selectedSwitcherIdx out of bound. Index has been reset to ${INITIAL_SWITHCER_INDEX}`
+            );
           }
         }),
         combineLatestWith(this.selectedSwitcherIdx$),
         tap(([switchers, selectedSwitcherIdx]) => {
-          if (!switchers.length) return;
-
           if (!this.isIndexValid(selectedSwitcherIdx, switchers)) {
-            console.warn('selectedSwitcherIdx out of bound');
+            this.logIndexValidationError('selectedSwitcherIdx out of bound');
             return;
           }
 
@@ -139,5 +141,9 @@ export class PrizmSwitcherComponent extends PrizmAbstractTestId implements Contr
 
   private isIndexValid(idx: number, switchers: PrizmSwitcherItem[]): boolean {
     return !!switchers[idx];
+  }
+
+  private logIndexValidationError(errorMsg: string) {
+    console.warn(errorMsg);
   }
 }

--- a/libs/components/src/lib/components/switcher/swithcer.const.ts
+++ b/libs/components/src/lib/components/switcher/swithcer.const.ts
@@ -1,0 +1,1 @@
+export const INITIAL_SWITHCER_INDEX = 0;

